Add a reader for EUMETSAT HSAF SEVIRI H60 and H63

albertbrotzer opened the pull request:
The H SAF product owners recently changed the output format of these products, and EUMETSAT requested support for the new format in EUMETView. 

 - [X] Closes #3282 
 - [X] Tests added  - not sure id this is sufficient
 - [X] Fully documented
 - [X] Add your name to `AUTHORS.md` if not there already

albertbrotzer committed:
Add initial HSAF precipitation product readers (h60b, h63, h90)

This commit introduces a reader enhancement for HSAF precipitation
products h60b, h63 and h90. The reader includes handling for externally
compressed (.gz) files by automatically unzipping them to a local
temporary location.

ameraner reviewed (changes requested):
Thanks for this! I would propose to actually have only one yaml file for this reader, and you can have different file types inside the same yaml (see e.g. the other PR here https://github.com/pytroll/satpy/pull/2793/files#diff-4ce2f1c002c4dae06506b9ad57f5a0504f278028002abc6263d2cd90ae50c8c6 ). This way it's more user friendly as there is only one reader for all HSAF products, and it's easier to maintain. 

Also, there are a few things to fix from the pre-commit checks.

ameraner commented on satpy/tests/enhancement_tests/test_hsaf.py:
```diff
@@ -0,0 +1,61 @@
+    # def test_detect_hsaf(self):
+    #     enh = Enhancer()
+    #     enh.add_sensor_enhancements(["hsaf"])
+    #     enh.enhancement_tree.print_tree()
```
Done

ameraner commented on satpy/readers/hsaf_nc.py:
```diff
@@ -0,0 +1,220 @@
+      uncompressed to the same directory and deleted on reader close.
+    - The reader relies on area definitions provided in the accompanying YAML
+      configuration file.
+    - Supports ``rain_rate`` and ``q_ind#`` datasets.
```
```suggestion
    - Supports ``rain_rate`` and ``q_ind`` datasets.
```
I guess the hashtag is a typo?

ameraner commented on satpy/readers/hsaf_nc.py:
```diff
@@ -0,0 +1,220 @@
+    - qind: quality index in percent
+
+    The data is on a geostationary projection grid.
+    This file handler is tested with H SAF h60, h63 & h40 data in NetCDF format.
```
```suggestion
    This file handler is tested with H SAF h60/h60b, h63 & h90 data in NetCDF format.
```

ameraner reviewed (changes requested):
Thank you for the updates! I took a closer looks and gave some more comments. The area definition assignments needs to be revised, otherwise only minor things.

ameraner commented on satpy/readers/hsaf_nc.py:
```diff
@@ -0,0 +1,220 @@
+        # Add metadata from YAML
+        for key, value in dataset_info.items():
+            if key == "area":
+                data.attrs[key] = _resolve_area(value)
```
(edited my comment above with a more elegant solution)

armelzer committed:
Consolidate HSAF reader configuration and address review remarks

This commit restructures and improves the reader following initial review
feedback.

Key changes:
- Unified all file_type definitions into a single YAML configuration file
- Simplified and cleaned up reader logic, removing redundant code
- Expanded and clarified docstrings across the reader implementation
- Improved handling of area definitions in the HSAF NetCDF reader
- Generalized dataset definitions so they can be reused by multiple file types

albertbrotzer committed:
Use generic_open for unified compressed/uncompressed file handling

Replaced the custom logic for handling externally zipped NetCDF files
with Satpy's generic_open helper, which transparently supports both
compressed and uncompressed inputs. This simplifies the reader
implementation and removes compression-detection code.

Because generic_open uses a context manager and closes the file as soon
as the reader exits the `with` block, the dataset must now be fully
loaded immediately. The reader therefore calls `.compute()` on the
xarray dataset to ensure all data is available after the context closes.

ameraner commented on satpy/readers/hsaf_nc.py:
```diff
@@ -0,0 +1,157 @@
+        self._area_name = None
+        self._lon_0 = None
+        with generic_open(filename, mode="rb", compression="infer") as fp:
+            self._nc_data = xr.open_dataset(fp, chunks="auto", engine="h5netcdf").compute()
```
I believe the `compute()` here is breaking the lazy reading of the data... If you remove it, does the reader still work, and we get dask arrays for the main variables? that should be the case..

ameraner commented on satpy/readers/hsaf_nc.py:
```diff
@@ -0,0 +1,157 @@
+        attributes = {
+            "filename": self.filename,
+            "start_time": self.start_time,
+            "end_time": self.end_time,
```
I believe these are not needed, as they're added automatically through the property methods

ameraner reviewed (commented):
Thanks for the updates! Looks good - only one point on the use of `compute`, and a minor point on the attributes

albertbrotzer committed:
Use generic_open context manager to ensure lazy-loaded NetCDF data remains accessible

- Replace the short-lived with generic_open(...) usage with manual
__enter__() / __exit__() handling

- Avoid loading the entire dataset via .compute(), preserving lazy
operation while preventing “seek of closed file” errors.

- Ensure the NetCDF file remains open for Dask-backed arrays until the
reader is destroyed.

- Add safe cleanup in __del__() using explicit context-manager exit and
catching expected cleanup-related exceptions.

albertbrotzer committed:
Remove start/end time from the attributes

...as they're added automatically through the property methods

ameraner reviewed (approved):
LGTM, thanks for taking the initiative to add this reader and all the updates!

ReadTheDocs is currently failing, but only due to a warning regarding not being able to fetch a numpy inventory due to HTTP error, so probably just a glitch and nothing related to the PR.

pnuu committed:
Merge branch 'main' into feature-hsaf-msg

mraspaud commented on satpy/readers/hsaf_nc.py:
```diff
@@ -0,0 +1,167 @@
+        self._lon_0 = None
+        # use generic_open contextmanager to handle compression transparently
+        self._cm = generic_open(filename, mode="rb", compression="infer")
+        # manually enter the context manager
```
```suggestion
```

mraspaud commented on satpy/readers/hsaf_nc.py:
```diff
@@ -0,0 +1,167 @@
+    def __del__(self):
+        """Instruct the context manager to clean up and close the file."""
+        try:
+            if hasattr(self, "_cm") and self._cm:
```
```suggestion
```

I don't think this is needed: `self._cm` will we be present and have a value as long as the class has been instantiated. If not, this `__del__` function will never be called anyway.

mraspaud commented on satpy/readers/hsaf_nc.py:
```diff
@@ -0,0 +1,167 @@
+        attributes = {
+            "filename": self.filename,
+            "spacecraft_name": platform_translate.get(self._nc_data.attrs["satellite_identifier"], "N/A"),
+            "platform_name": platform_translate.get(self._nc_data.attrs["satellite_identifier"], "N/A"),
```
is `N/A` an acceptable value for these attributes? Does it ever happen that `satellite_identifier` is not defined?
Also, why do we need both `spacecraft_name` and `platform_name`?

mraspaud commented on satpy/readers/hsaf_nc.py:
```diff
@@ -0,0 +1,167 @@
+
+        data = self._standarize_dims(data)
+
+        # Handle missing values
```
```suggestion
```

I think the code is self explanatory.

mraspaud commented on satpy/readers/hsaf_nc.py:
```diff
@@ -0,0 +1,167 @@
+        self._area_name = self.filetype_info["area"]
+        self._lon_0 = float(self._nc_data.attrs["sub_satellite_longitude"].rstrip("f"))
+
+        # Add global attributes which are shared across datasets
```
```suggestion
```

mraspaud commented on satpy/tests/reader_tests/test_hsaf_nc.py:
```diff
@@ -0,0 +1,186 @@
+    def test_reader_creation(self, file_type, expected_loadables):
+        """Test that the reader can create file handlers."""
+        with mock.patch("satpy.readers.hsaf_nc.generic_open") as mock_generic_open, \
+                mock.patch("satpy.readers.hsaf_nc.xr.open_dataset") as mock_open_dataset:
```
We have now as policy to try to create a stub file in `tmp_path` instead of mocking. I think here, it's a simple matter of writing the xarray dataset to disk and use that as input file. With the added benefit that you can (g/b)zip it to test the decompression...

mraspaud commented on satpy/tests/reader_tests/test_hsaf_nc.py:
```diff
@@ -0,0 +1,186 @@
+        """Test the reader with a real HSAF NetCDF file."""
+        # Select files
+        loadables = file_type["reader"].select_files_from_pathnames([file_type["real_file"]])
+        assert loadables, "No loadables found for the real file"
```
when are these real files available? are they easily downloadable from somewhere so that the user can test this?
Otherwise, I think it's as good to create stub files that mimic these files closely.

mraspaud reviewed (changes requested):
Thanks a lot for the work on supporting these HSAF products!
I especially appreciate the clarity of the code.

I have some comments, questions and suggestions inline, but I think this is almost ready :)

armelzer committed:
Move definition of timedelta value for end_time calculation

This will make it possible to use this reader for other H-SAF products (e.g. H40/H40B) in the future,
provided that the corresponding file type is configured.

armelzer committed:
Clean up unnecessary comments and code

armelzer committed:
Create a stub files instead of mocking

...so that it complies with the guidelines for the use of stub files.

mraspaud commented on satpy/etc/enhancements/hsaf.yaml:
```diff
@@ -0,0 +1,39 @@
+                20,
+                30,
+                50
+              ]
```
I'm no expert here, @pnuu @ameraner what do you say?

armelzer committed:
Introduce separate standard_names for `acc_rr` and `rr`

...so the defined enhancement will only be applied to `rr`.

armelzer committed:
Fix test execution

With this commit the tear down method of the `TestHSAFNCReader` will
be correctly called and the test for the enhancement will no longer fail.

ameraner committed:
Merge branch 'main' into feature-hsaf-msg

mraspaud reviewed (approved):
LGTM
